File: json-performance/performance-object-parser/src/main/java/com/steelzack/testing/hierarchy/CodeResolver.java
package com.steelzack.testing.hierarchy;

import java.util.HashMap;
import java.util.Map;

public class CodeResolver implements CodeHierarchy {

	private Map<String, ParentCode> allMaps = new HashMap<String, ParentCode>();

	@Override
	public boolean isChildOf(String aCode, String bCode) {
		Code codeA = allMaps.get(aCode);
		Code codeB = allMaps.get(bCode);

		boolean isChild = false;
		if (codeA.getParent().getName().equals(codeB.getName())) {
			isChild = true;
		}
		return isChild;
	}

	@Override
	public boolean isParentOf(String aCode, String bCode) {
		Code codeA = allMaps.get(aCode);
		Code codeB = allMaps.get(bCode);

		boolean isParent = false;
		if (codeB.getParent().getName().equals(codeA.getName())) {
			isParent = true;
		}
		return isParent;
	}

	@Override
	public boolean inHierarchyOf(String aCode, String bCode) {
		return isChildOf(aCode, bCode) || isParentOf(aCode, bCode);
	}

	public static void main(String[] args) {
		String[][] matrix = new String[][] { { "A", "A", "It's A" },
				{ "B", "A", "It's B" }, { "C", "A", "It's C" },
				{ "D", "C", "It's D" } };

		CodeResolver resolver = new CodeResolver();
		resolver.proccessMatrix(matrix);

	}

	public void proccessMatrix(String[][] matrix) {
		for (int i = 0; i < matrix.length; i++) {
			String parentString = matrix[i][1];
			String childString = matrix[i][0];
			String coment = matrix[i][2];

			ParentCode parent = (ParentCode) allMaps.get(parentString);
			if (parent == null) {
				parent = new ParentCode(parentString, coment);
			}

			ParentCode code = (ParentCode) allMaps.get(parentString);

			if (code == null) {
				code = new ParentCode(parent, childString);
			}
			
			parent.addChild(childString, code);
			allMaps.put(parentString, parent);
			allMaps.put(childString, code);
		}
		
		System.out.println(isParentOf("A", "B"));
		System.out.println(inHierarchyOf("A", "D"));
	}

}

File: jeorg-kotlin-apps/performance-parsers/json-performance/performance-object-parser/src/main/kotlin/org/jesperancinha/testing/hierarchy/CodeResolver.kt
package org.jesperancinha.testing.hierarchy

class CodeResolver : CodeHierarchy {
    private val allMaps: MutableMap<String, ParentCode> = HashMap()
    override fun isChildOf(aCode: String, bCode: String): Boolean {
        val codeA: Code? = allMaps[aCode]
        val codeB: Code? = allMaps[bCode]
        var isChild = false
        if (codeA!!.parent!!.name == codeB!!.name) {
            isChild = true
        }
        return isChild
    }

    override fun isParentOf(aCode: String, bCode: String): Boolean {
        val codeA: Code? = allMaps[aCode]
        val codeB: Code? = allMaps[bCode]
        var isParent = false
        if (codeB!!.parent!!.name == codeA!!.name) {
            isParent = true
        }
        return isParent
    }

    override fun inHierarchyOf(aCode: String, bCode: String): Boolean {
        return isChildOf(aCode, bCode) || isParentOf(aCode, bCode)
    }

    fun proccessMatrix(matrix: Array<Array<String>>) {
        for (i in matrix.indices) {
            val parentString = matrix[i][1]
            val childString = matrix[i][0]
            val coment = matrix[i][2]
            var parent = allMaps[parentString]
            if (parent == null) {
                parent = ParentCode(parentString, coment)
            }
            var code = allMaps[parentString]
            if (code == null) {
                code = ParentCode(parent, childString)
            }
            parent.addChild(childString, code)
            allMaps[parentString] = parent
            allMaps[childString] = code
        }
        println(isParentOf("A", "B"))
        println(inHierarchyOf("A", "D"))
    }

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            val matrix = arrayOf(
                arrayOf("A", "A", "It's A"),
                arrayOf("B", "A", "It's B"),
                arrayOf("C", "A", "It's C"),
                arrayOf("D", "C", "It's D")
            )
            val resolver = CodeResolver()
            resolver.proccessMatrix(matrix)
        }
    }
}